build/BUILD

```diff
@@ -55,10 +53,6 @@ DOCKERIZED_BINARIES = {
     architectures = SERVER_PLATFORMS["linux"],
     base = meta["base"],
     cmd = ["/usr/bin/" + binary],
-    debs = select(for_platforms(
-        for_server = ["//build/debs:%s-{ARCH}.deb" % binary],
-        only_os = "linux",
-    )),
```

Comment:
i don't think this `multi_arch_container()` container declaration works currently.

looking at:
https://github.com/kubernetes/kubernetes/blob/master/build/container.bzl#L44

i don't see the `cmd` argument in the list...or maybe i just can't read bazel.
if i'm correct and this declaration doesn't work as expected, wondering why are we not seeing errors.

do we have an active maintainer from /build who can help here?

Reply:
...perhaps?
/assign @cblecker @BenTheElder @dims

Reply:
i mean regardless of cmd, these binaries are getting into the images by way of passing the debs, so that's problematic.

Reply:
`debs` is also not an accepted arg.

Reply:
it is possible to implement some other way, but we can't just remove them, this is why CI is failing

Reply:
@neolit123 these are kwargs, see https://github.com/bazelbuild/rules_docker#container_image-1

skylark is essentially a restricted subset of python with less mutation and builtins

Reply:
i see, so these flags are passed to the underlying `container_image()`

> debs
> List of files, optional
> Deprecated: A list of debian packages that will be extracted in the Docker image. Note that this doesn't actually install the packages. Installation needs apt or apt-get which need to be executed within a running container which container_image can't do.

so it seems safe to just remove the `debs` arg.

Reply:
to elaborate:
`multi_arch_container` accepts named args [and kwargs](https://github.com/kubernetes/kubernetes/blob/6ecdd22978783ea51efc34f56f13eb4d9baa446a/build/container.bzl#L52), the kwargs [are passed](https://github.com/kubernetes/kubernetes/blob/6ecdd22978783ea51efc34f56f13eb4d9baa446a/build/container.bzl#L61) on to a `container_image` invocation, which accepts `cmd` and `deb` as args (see link above)

the debs are installed to the image, and cmd is image `CMD`.

https://docs.bazel.build/versions/master/skylark/language.html

https://github.com/google/skylark/blob/master/doc/spec.md#functions subheading "Keyword-variadic functions"

Reply:
> so it seems safe to just remove the debs arg.

No, because we don't have real install scripts, our debs are pretty much just tarballs of files.
The files need to be in the images .... otherwise these images are useless

Reply:
trying to figure out the logic here:

- there is a target list of binaries:
DOCKERIZED_BINARIES (control-plane & kube-proxy)
- we iterate trough that list and build container images using `multi_arch_container` from the Bazel docker rules
- the function `multi_arch_container()` accepts the optional `debs`:
> A list of debian packages that will be installed in the Docker image
- for `debs` we pass the Debian packages for kubeadm and the kubelet

the question here is why the control-plane and kube-proxy images need the kubelet and kubeadm binaries installed? i don't see why should they.

Reply:
placing debs in the images seems odd.

i think a combination of `files` + `directory` should work here, maybe:

```
[multi_arch_container(
    name = binary,
    architectures = SERVER_PLATFORMS["linux"],
    base = meta["base"],
    cmd = ["/usr/bin/" + binary],
    directory =  "/usr/bin/",
    files =  ["//build/%s-{ARCH}" % binary], # <--- syntax / location?
    # Since the multi_arch_container macro replaces the {ARCH} format string,
    # we need to escape the stamping vars.
    docker_push_tags = ["{{STABLE_DOCKER_PUSH_REGISTRY}}/%s-{ARCH}:{{STABLE_DOCKER_TAG}}" % binary],
    docker_tags = ["{{STABLE_DOCKER_REGISTRY}}/%s-{ARCH}:{{STABLE_DOCKER_TAG}}" % binary],
    stamp = True,
    symlinks = {
        "/usr/bin/" + binary: "/usr/bin/" + "%s-{ARCH}" % binary,
    },
    tags = ["manual"],
    visibility = ["//visibility:private"],
) for binary, meta in DOCKERIZED_BINARIES.items()]
```

Reply:
The binary's Deb is passed, not all of them.

We can accomplish it with something like above, but we can't just remove the deb bit without writing the replacement 🙃

Reply:
I think @ixdy wrote this as a way to make the images without duplicating the details of where binaries go etc. It used to not be deprecated 🤷‍♂️

Reply:
I think using debs for these images predates even my work on this. (I think that was @mikedanese?)

But as far as I can recall, it was as @BenTheElder said - this prevented us from needing to list all dependencies twice.

Reply:
(also, the docker/container rules were far less mature when we first implemented this, and we may have only been able to add files in the right places with the right permissions using debs.)

Reply:
> We can accomplish it with something like above, but we can't just remove the deb bit without writing the replacement upside_down_face

ok, seems to me this is doable but this part seems incomplete:

```
files =  ["//cmd/" + binary], # <--- FIXME; where is this built?
```

maybe bazel seeing `//cmd/kube-apiserver` will trigger the build for that, maybe not.

Reply:
> (also, the docker/container rules were far less mature when we first implemented this, and we may have only been able to add files in the right places with the right permissions using debs.)

ok, i see. so it seems the debs here are used as a workaround at least for the file permissions.

so maybe we can do the following:
- stop building debs for everything other than `kube-<control-plane-component>` and `kube-proxy`
- continue using the DEB workaround for those
- stop building RPMs completely
- stop pushing DEBs and RPMs to the `kubernetes-release-dev` GCS.

Reply:
IIRC, tars would work fine pkg_deb was used instead of pkg_tar because we needed debs anyway.  deb_data (maybe poorly named at this point) produces a tar that would probably work as is:

https://github.com/kubernetes/repo-infra/blob/1bf7b575499e998e792c29849d0ad1d0097298b3/defs/deb.bzl#L47-L66

Just pass the result of that to tars instead of debs, and I would expect everything to click.